terraform/node_data_refresh.go

```diff
@@ -54,22 +91,18 @@ func (n *NodeRefreshableDataResource) DynamicExpand(ctx EvalContext) (*Graph, er
 	// if we're transitioning whether "count" is set at all.
 	fixResourceCountSetTransition(ctx, n.ResourceAddr(), count != -1)
 
-	var instanceAddrs []addrs.AbsResourceInstance
-
 	// Inform our instance expander about our expansion results above,
 	// and then use it to calculate the instance addresses we'll expand for.
 	expander := ctx.InstanceExpander()
-	for _, path := range expander.ExpandModule(n.Addr.Module) {
-		switch {
-		case count >= 0:
-			expander.SetResourceCount(path, n.ResourceAddr().Resource, count)
-		case forEachMap != nil:
-			expander.SetResourceForEach(path, n.ResourceAddr().Resource, forEachMap)
-		default:
-			expander.SetResourceSingle(path, n.ResourceAddr().Resource)
-		}
-		instanceAddrs = append(instanceAddrs, expander.ExpandResource(n.ResourceAddr().Absolute(path))...)
+	switch {
+	case count >= 0:
+		expander.SetResourceCount(n.Addr.Module, n.Addr.Resource, count)
+	case forEachMap != nil:
+		expander.SetResourceForEach(n.Addr.Module, n.Addr.Resource, forEachMap)
+	default:
+		expander.SetResourceSingle(n.Addr.Module, n.Addr.Resource)
```

Comment:
Question: what is the expected behavior when `SetResourceSingle` is called when `count = 0` ?

Reply:
Ah! I found the answer in a comment just below this:
> // -1 signals "count not set"

Reply:
yup, `count = 0` is still in list-most, just with 0 instances.